examples/cpp/pyperf/PyPerfNativeStackTrace.cc

```diff
@@ -0,0 +1,152 @@
+  my_accessors.access_reg = NativeStackTrace::access_reg;
+
+  unw_addr_space_t as = unw_create_addr_space(&my_accessors, 0);
+  void *upt = _UPT_create(pid);
+  if (!upt) {
+    logInfo(1, "Failed to _UPT_create\n");
+    this->symbols.push_back(std::string("[Error]"));
+    errors++;
+    return;
+  }
+
+  unw_cursor_t cursor;
+  int res = unw_init_remote(&cursor, as, upt);
+  if (res) {
+    logInfo(1, "Failed to unw_init_remote: %s\n", unw_strerror(res));
+    this->symbols.push_back(std::string("[Error]"));
+    errors++;
+    return;
+  }
+
+  do {
+    unw_word_t offset;
+    char sym[256];
+
+    if (unw_get_proc_name(&cursor, sym, sizeof(sym), &offset) == 0) {
+      this->symbols.push_back(std::string(sym));
+    } else {
+      unw_word_t ip;
+      unw_get_reg(&cursor, UNW_REG_IP, &ip);
+      logInfo(2,
+              "0x%lx -- error: unable to obtain symbol name for this frame "
+              "(SP=0x%lx)\n",
+              ip, NativeStackTrace::sp);
+      this->symbols.push_back(std::string("(missing)"));
+      errors++;
+      break;
+    }
+
+    if (memcmp(sym, "_PyEval_EvalFrameDefault",
+                sizeof("_PyEval_EvalFrameDefault")) == 0 ||
+        memcmp(sym, "PyEval_EvalFrameEx", sizeof("PyEval_EvalFrameEx")) == 0)
+        {
+      break;
+    }
+  } while (unw_step(&cursor) > 0);
+
+  _UPT_destroy(upt);
+  return;
+}
+
+int NativeStackTrace::access_reg(unw_addr_space_t as, unw_regnum_t regnum,
+                                 unw_word_t *valp, int write, void *arg) {
+  if (regnum == UNW_REG_SP) {
+    if (write) {
+      logInfo(2, "Libunwind attempts to write to SP\n");
+      return -UNW_EINVAL;
+    }
+
+    *valp = NativeStackTrace::sp;
+    return 0;
+  }
+  else if (regnum == UNW_REG_IP) {
+    if (write) {
+      logInfo(2, "Libunwind attempts to write to IP\n");
+      return -UNW_EINVAL;
+    }
+
+    *valp = NativeStackTrace::ip;
+    return 0;
+  }
+  else {
+    logInfo(2, "Libunwind attempts to %s regnum %d\n", write ? "write" : "read", regnum);
+    return -UNW_EBADREG;
+  }
+}
+
+int NativeStackTrace::access_mem(unw_addr_space_t as, unw_word_t addr,
+                                 unw_word_t *valp, int write, void *arg) {
+  if (write) {
+    logInfo(2, "Libunwind mem write attempt\n");
+    return -UNW_EINVAL;
+  }
+
+  const unw_word_t stack_start = NativeStackTrace::sp & ~(getpagesize() - 1);
+  const unw_word_t stack_end = stack_start + NativeStackTrace::stack_len;
+
+  if (addr >= NativeStackTrace::sp && addr < stack_end) {
+    memcpy(valp, &stack[addr - stack_start], sizeof(*valp));
+    return 0;
+  } else if ((addr >= stack_end && addr < stack_end + getpagesize() * 2) ||
+             (addr >= stack_start - getpagesize() && addr < NativeStackTrace::sp)) {
+    // Memory accesses around the pages we copied are assumed to be accesses to the
+    // stack that we shouldn't allow
+    logInfo(2, "Libunwind failed attempt to access stack at 0x%lx (SP=0x%lx)\n", addr,
```

Comment:
1. Are these logs actually written (I ran PyPerf from your branch and I didn't see any)
2. I assume they are not; those that are real errors - we should make sure they get written when gProfiler runs PyPerf, because that'll allow us to debug problems. Now, this particular one is not a real error, I think? libunwind will *always* try to access off limits (because usually 1-2-3 pages are not enough) so this will get written every time. If that's the case, we probably don't want this print.

Reply:
In some cases such access indicates a problem. As you wrote, usually a page or two is enough to get to the beginning of a python function. This print is useful to distinguish between an actual error and a valid access attempt out of the scope of the stack that was copied (mostly this print will appear when there is no error, when the address is greater than SP).
Maybe I didn't understand the usage policy of `logInfo` in the project. Maybe I should have used log level `3` to print this kind of debug info?

Reply:
Not sure how spammy this is, but I think level 2 fits here. The wording can be slightly confusing as the reason for the failure is unclear (invalid address? valid address but page not copied? or something else?).

Reply:
I clarified the error message